lgc/builder/ImageBuilder.cpp

```diff
@@ -1257,12 +1257,37 @@ Value *ImageBuilder::CreateImageAtomicCommon(unsigned atomicOp, unsigned dim, un
 Value *ImageBuilder::CreateImageQueryLevels(unsigned dim, unsigned flags, Value *imageDesc, const Twine &instName) {
   imageDesc = patchInvalidImageDescriptor(imageDesc);
   dim = dim == DimCubeArray ? DimCube : dim;
-  Value *zero = getInt32(0);
-  Instruction *resInfo = CreateIntrinsic(ImageGetResInfoIntrinsicTable[dim], {getFloatTy(), getInt32Ty()},
-                                         {getInt32(8), UndefValue::get(getInt32Ty()), imageDesc, zero, zero});
-  if (flags & ImageFlagNonUniformImage)
+
+  if (flags & ImageFlagNonUniformImage) {
+    Value *zero = getInt32(0);
+    Instruction *resInfo = CreateIntrinsic(ImageGetResInfoIntrinsicTable[dim], {getFloatTy(), getInt32Ty()},
+                                           {getInt32(8), UndefValue::get(getInt32Ty()), imageDesc, zero, zero});
     resInfo = createWaterfallLoop(resInfo, 2);
-  return CreateBitCast(resInfo, getInt32Ty(), instName);
+    return CreateBitCast(resInfo, getInt32Ty(), instName);
+  } else {
+    Value *numMipLevel;
+    if (dim == Dim2DMsaa || dim == Dim2DArrayMsaa)
+      numMipLevel = getInt32(1);
+    else {
+      // Extract LAST_LEVEL (SQ_IMG_RSRC_WORD3, [19:16])
+      Value *descWord3 = CreateExtractElement(imageDesc, 3);
+      Value *lastLevel = CreateIntrinsic(Intrinsic::amdgcn_ubfe, getInt32Ty(), {descWord3, getInt32(16), getInt32(4)});
```

Comment:
This all needs abstracting out to account for different hardware having different descriptor layouts, right?

Reply:
Yeah, different gfxips may have different descriptor layouts.
I just follow the design of "preprocessIntegerImageGather()" or "CreateImageQuerySamples()".
But they are much simpler.
So what do you suggest here? Shall I put those descriptor layouts related code into functions?
Like class "YCbCrAddressHandler" for "YCbCrConverter"?

Reply:
preprocessIntegerImageGather gets away with it because it is only used for pre-gfx9.

I see what you mean that CreateImageQuerySamples has the same style you're using, only it is smaller.

Perhaps we could have separate functions for getImageDescLastLevel, getImageDescBaseLevel, etc, some of which are used in both CreateImageQuerySamples and CreateImageQueryLevels. Then those are the places we need to insert gfx-specific code later on if future hardware changes the descriptor layout.

What do others think?

Reply:
You should use `SqImgRsrcRegHandler` for this.

Reply:
OK, there is already a SqImgRsrcRegHandler. 
But it seems that some of fields (mip levels) are missing in SqImgRsrcRegHandler; and it does not support NullDescriptor case.
I guess I will have to fix them and then refine the change with SqImgRsrcRegHandler.